pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go

```diff
@@ -0,0 +1,938 @@
+// remainFree = allowed cpu - reserved cpu - sum(requests of containers in each dynamicPool)
+func (p *dynamicPools) calculateAllPoolRequests() (map[*DynamicPool]int, int) {
+	requestCpu := make(map[*DynamicPool]int)
+	remainFree := p.allowed.Difference(p.reserved).Size()
+	for _, dp := range p.dynamicPools {
+		if dp.Def.Name == "reserved" {
+			continue
+		}
+		requestCpu[dp] = (p.requestedMinMilliCpus(dp) + 999) / 1000
+		remainFree -= requestCpu[dp]
+		log.Info("dyanmic pool %s request cpu %d", dp, requestCpu[dp])
+	}
+	log.Info("sum remain free cpu %d", remainFree)
+	return requestCpu, remainFree
+}
+
+func (p *dynamicPools) containerPinPool(dp *DynamicPool) {
+	dp.Mems = p.closestMems(dp.Cpus)
+	for _, cID := range dp.ContainerIDs() {
+		if c, ok := p.cch.LookupContainer(cID); ok {
+			p.pinCpuMem(c, dp.Cpus, dp.Mems)
+		}
+	}
+}
+
+// updatePoolCpuset updates the cpuset of the dynamicPools.
+func (p *dynamicPools) updatePoolCpuset() error {
+	requestCpu, remainFree := p.calculateAllPoolRequests()
+	weight, sumWeight, err := p.calculateAllPoolWeights()
+	if err != nil {
+		return err
+	}
+
+	if remainFree >= 1 {
+		usedCpu := 0
+		// Ensure that there is at least one cpu in the shared dynamicPool.
+		for _, dp := range p.dynamicPools {
+			if dp.Def.Name == "shared" && sumWeight != 0 {
+				addCpu := remainFree * weight[dp] / sumWeight
+				if requestCpu[dp]+addCpu < 1 {
+					requestCpu[dp] = 1
+					remainFree -= 1
+				}
+			}
+		}
+		for _, dp := range p.dynamicPools {
+			if dp.Def.Name == "reserved" {
+				continue
+			}
+			if sumWeight != 0 {
+				addCpu := remainFree * weight[dp] / sumWeight
+				requestCpu[dp] += addCpu
+				usedCpu += addCpu
+			}
+			log.Info("dyanmic pool %s new request cpu %d, remain free cpu %d", dp, requestCpu[dp], remainFree-usedCpu)
+		}
+		if usedCpu < remainFree {
+			// If there is still cpus, give the dynamicPool with the highest cpu utilization.
+			tmp := p.dynamicPools[1]
+			for _, dp := range p.dynamicPools {
+				if dp.Def.Name == "reserved" {
+					continue
+				}
+				if weight[dp] > weight[tmp] {
+					tmp = dp
+				}
+			}
+			requestCpu[tmp] += (remainFree - usedCpu)
+			log.Info("dyanmic pool %s new request cpu %d, remain free cpu %d", tmp, requestCpu[tmp], 0)
+		}
+	}
+
+	// If the number of newly allocated CPUs is the same as the number of existing CPUs in the pool,
+	// it means that there is no need to re-allocate
+	isNeedReallocate := false
+	for _, dp := range p.dynamicPools {
+		if dp.Def.Name == "reserved" {
+			continue
+		}
+		if dp.Cpus.Size() != requestCpu[dp] {
+			isNeedReallocate = true
+			break
+		}
+	}
+	if !isNeedReallocate {
+		log.Info("The number of CPUs required by the pools is the same as the number of CPUs already in the pools, so there is no need to reallocate.")
+		for _, dp := range p.dynamicPools {
+			p.containerPinPool(dp)
+		}
+		return nil
+	}
+
+	for _, dp := range p.dynamicPools {
+		if dp.Def.Name == "reserved" {
```

Comment:
Reallocation of CPUs can severely harm performance of a workload. If the workload has been running on CPU package0 (socket0) and is suddenly pinned to CPU cores on package1 (socket1), then suddenly memory accesses become very expensive.

In order to avoid unnecessary CPU reallocations, I would suggest:
- First release only the necessary amount of CPUs from a pool (current number of CPUs minus requestCpu[dp], if this is a positive number).
- Then allocate CPUs only to those pools that really need them.

The current balloons policy implements topology-aware CPU allocation, that is, it allocates CPU cores that are topologically as close to the currently allocated CPUs as possible. This is a bigger enhancement and can be left to be done in future PRs.

Reply:
Thanks for your suggestions! 
This is indeed where the dynamic-pools policy needs to be improved. We will seriously consider your suggestions, and we will improve the performance of dynamic-pools policy in the future.